### PR TITLE
Generalize storage formats in file metastore

### PR DESCRIPTION
**Pull request by 7c00**
In this PR, we changed the storage format class from `HiveStorageFormat` to `StorageFormat`, in order to make it possible to create tables and partitions whose stoarge formats are not listed in the  `HiveStorageFormat`. For example, a Hudi COW table is stored in meatastore with the input format as `HoodieParquetInputFormat`.

This PR is extracted from the some legacy version of https://github.com/prestodb/presto/pull/17149

Test plan - Unit tests.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by 7c00** [comment]
ping @arunthirupathi

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java
Comment (arunthirupathi) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java:91]: Looks like this is always present, why have the storageFormat as optional then ?

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/TableMetadata.java
Comment (arunthirupathi) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/TableMetadata.java:111]: same comment.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java
Comment (7c00) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java]: Right. The optional here is unnecessary. Have pruned.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/TableMetadata.java
Comment (7c00) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/TableMetadata.java]: Fixed as well.

**Review by arunthirupathi — CHANGES_REQUESTED** [review]
This change is backward incompatible. 

JsonCreator is an annotation used for creating class dynamically from JSON. This change modifies the type. When existing JSON blobs are de-serialized they will fail due to the type change. 

Also looks like no-one wrote the tests for it originally for serialization so none of the tests broke.

Can you please investigate how to make these changes without breaking the existing assumptions ? Generally you have to introduce a new property and deprecate the existing property and optionally after few months remove the deprecated property.

**Comment by 7c00** [comment]
@arunthirupathi Thanks for your comments. 

I have updated the code to make the legacy json (that is serialized from the previous type definition) readable. The legacy constructors and methods are keeping existing but marked as deprecated. Could you please take a second review?

**Comment by arunthirupathi** [comment]
Thanks for making this backward compatible, but I don't see tests that capture the old json and verify they are successfully deserialized.

I want to see the following things in this PR.

1. Checkout master, serialize some examples (with optional.absent and optional.present with different values).
2. Create tests that instantiates the new objects from this examples.
3. Make your code changes.
4. The tests should still pass for the old values.

Apologies for asking you to create tests, but this code is widely used and we need to be careful. This has a potential to break almost all our users and it is sad that the existing code did not capture the tests.

**Comment by 7c00** [comment]
It makes sense to add tests to ensure good compatibility! After refactoring the code, I split it into two commits: the first one does json round trip tests and encodes the instances of legacy classes to files; the second one adds tests to decode the file by new classes. Could you take a look to check if this is ok?

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java
Comment (arunthirupathi) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java:110]: formatting: one parameter per line

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/StorageFormatCompatDeserializer.java
Comment (arunthirupathi) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/StorageFormatCompatDeserializer.java:36]: Please add a comment here, prior to version 0.271 this was HiveStorageFormat and this class is to ensure back ward compatibility.

**Review by arunthirupathi — APPROVED** [review]
Left some minor nits, once addressed I will merge this in.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java
Comment (7c00) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java]: Fixed.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/StorageFormatCompatDeserializer.java
Comment (7c00) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/StorageFormatCompatDeserializer.java:38]: Added.

**Comment by 7c00** [comment]
@arunthirupathi The PR is ready to merge. Could you take a look when you have time? Thanks in advance.

**Pull request merged**
